tests/test_training/test_trainer.py

```diff
@@ -0,0 +1,226 @@
+from mrmustard.training import Optimizer
+from mrmustard.training.trainer import map_trainer, train_device, update_pop
+
+
+@pytest.fixture(scope="function")
+def wrappers():
+    """Dummy wrappers tested."""
+
+    def make_circ(x=0.0, return_type=None):
+        circ = Ggate(num_modes=1, symplectic_trainable=True) >> Dgate(
+            x=x, x_trainable=True, y_trainable=True
+        )
+        return (
+            [circ] if return_type == "list" else {"circ": circ} if return_type == "dict" else circ
+        )
+
+    def cost_fn(circ=make_circ(0.1), y_targ=0.0):
+        target = Gaussian(1) >> Dgate(-0.1, y_targ)
+        s = Vacuum(1) >> circ
+        return -fidelity(s, target)
+
+    return make_circ, cost_fn
+
+
+@pytest.mark.parametrize(
+    "tasks", [5, [{"y_targ": 0.1}, {"y_targ": -0.2}], {"c0": {}, "c1": {"y_targ": 0.07}}]
+)
+@pytest.mark.parametrize("seed", [None, 42])
+def test_circ_cost(wrappers, tasks, seed):  # pylint: disable=redefined-outer-name
+    """Test distributed cost calculations."""
+    has_seed = isinstance(seed, int)
+    _, cost_fn = wrappers
+    results = map_trainer(
+        cost_fn=cost_fn,
+        tasks=tasks,
+        **({"SEED": seed} if has_seed else {}),
+    )
+
+    if isinstance(tasks, dict):
+        assert set(results.keys()) == set(tasks.keys())
+        results = list(results.values())
+    assert all(r["optimizer"] is None for r in results)
+    assert all(r["device"] == [] for r in results)
+    if has_seed and isinstance(tasks, int):
+        assert len(set(r["cost"] for r in results)) == 1
+    else:
+        assert (
+            len(set(r["cost"] for r in results))
+            >= (tasks if isinstance(tasks, int) else len(tasks)) - 1
+        )
+
+
+@pytest.mark.parametrize(
+    "tasks", [[{"x": 0.1}, {"y_targ": 0.2}], {"c0": {}, "c1": {"euclidean_lr": 0.02, "HBAR": 1.0}}]
+)
+@pytest.mark.parametrize(
+    "return_type",
+    [None, "dict"],
+)
+def test_circ_optimize(wrappers, tasks, return_type):  # pylint: disable=redefined-outer-name
+    """Test distributed optimizations."""
+    max_steps = 15
+    make_circ, cost_fn = wrappers
+    results = map_trainer(
+        cost_fn=cost_fn,
+        device_factory=make_circ,
+        tasks=tasks,
+        max_steps=max_steps,
+        symplectic_lr=0.05,
+        return_type=return_type,
+    )
+
+    if isinstance(tasks, dict):
+        assert set(results.keys()) == set(tasks.keys())
+        results = list(results.values())
+    assert (
+        len(set(r["cost"] for r in results))
+        >= (tasks if isinstance(tasks, int) else len(tasks)) - 1
+    )
+    assert all(isinstance(r["optimizer"], Optimizer) for r in results)
+    assert all((r["optimizer"].opt_history) for r in results)
+
+    # Check if optimization history is actually decreasing.
+    opt_history = np.array(results[0]["optimizer"].opt_history)
+    assert len(opt_history) == max_steps + 1
+    assert opt_history[0] - opt_history[-1] > 1e-6
```

Comment:
note that as a quirk of opt_history, the first value is fixed 0 (in case this affects this test)